### PR TITLE
refactor(condo): fix race-condition

### PR DESCRIPTION
**Pull request by toplenboren**
Upgrades the L1-cache

### CODE REVIEW

File: packages/keystone/cache.js [hunk 1]
```diff
@@ -77,10 +91,13 @@ const patchMutation = (mutationContext, mutation, isUpdateMutation, cache) => {
 
 const patchQuery = (queryContext, query, cache) => {
     return async function (args, context, gqlName, info, from) {
+        requests++
```
Comment (SavelevMatthew) [inline packages/keystone/cache.js:94]: Why not inside class?

File: packages/keystone/cache.js [hunk 2]
```diff
@@ -29,6 +29,13 @@
 const express = require('express')
 const { get, set } = require('lodash')
 
+const ENABLE_CACHE_LOGGING = true
```
Comment (SavelevMatthew) [inline packages/keystone/cache.js:32]: Why not from .env? Is it WIP? When why not tagged?

**Review by SavelevMatthew — DISMISSED** [review]

File: packages/keystone/cache.js (hunk 2 above)
Comment (toplenboren) [reply packages/keystone/cache.js:32]: yep, this is wip. Forgot to tag. Gonna dismiss and re-request your review once ready

File: packages/keystone/cache.js [hunk 3]
```diff
@@ -44,6 +57,11 @@ class KeystoneCacheMiddleware {
             const requestId = get(req, ['headers', 'x-request-id'])
             res.on('close', () => {
                 if (requestId) {
+
+                    if (ENABLE_CACHE_LOGGING) {
+                        logger.info( { msg: `DELETE_FROM_MIDDLEWARE: ${requestId}` })
```
Comment (pahaz) [inline packages/keystone/cache.js:62]:
```suggestion
                        logger.info(`DELETE_FROM_MIDDLEWARE: ${requestId}`)
```

File: packages/keystone/cache.js [hunk 4]
```diff
@@ -61,68 +79,103 @@ const getRequestIdFromContext = (context) => {
...
+        cacheMiddleware.requests++
+
         let key = null
         const requestId = getRequestIdFromContext(context)
 
         if (requestId) {
+            cacheMiddleware.hasRequestId++
+
+            if (!args.first) {
+                args.first = 100
```
Comment (pahaz) [inline packages/keystone/cache.js:107]: O_o ?

File: packages/keystone/cache.js [hunk 5]
```diff
@@ -61,68 +79,103 @@ const getRequestIdFromContext = (context) => {
...
+                if (ENABLE_CACHE_LOGGING) {
+                    logger.info({ msg: `HIT: ${requestId}\r\n${gqlName} ${JSON.stringify(args)}\r\nScore: ${cacheMiddleware.hits}/${cacheMiddleware.requests} Request ids: ${cacheMiddleware.hasRequestId}/${requestId} ` })
+                }
+
+                return await cacheMiddleware.cache[requestId][key]
+            }
+        }
+        else {
+            if (ENABLE_CACHE_LOGGING) {
+                logger.info({ msg: `WARNING: No request ID\r\n${gqlName} ${JSON.stringify(args)}` })
```
Comment (pahaz) [inline packages/keystone/cache.js:139]:
```suggestion
                logger.info(`WARNING: No request ID\r\n${gqlName} ${JSON.stringify(args)}`)
```

File: packages/keystone/cache.js [hunk 6]
```diff
@@ -29,12 +29,25 @@
...
 
     constructor (keystone) {
         this.cache = {}
 
-        initCache(keystone, this.cache)
+        this.hits = 0
+        this.requests = 0
+        this.hasRequestId = 0
+
+        initCache(keystone, this)
```
Comment (pahaz) [inline packages/keystone/cache.js:50]: Why not to create just StateObject instead of this?

File: packages/keystone/cache.js [hunk 7]
```diff
@@ -61,68 +79,103 @@ const getRequestIdFromContext = (context) => {
...
+                cacheMiddleware.hits++
+
+                if (ENABLE_CACHE_LOGGING) {
+                    logger.info({ msg: `HIT: ${requestId}\r\n${gqlName} ${JSON.stringify(args)}\r\nScore: ${cacheMiddleware.hits}/${cacheMiddleware.requests} Request ids: ${cacheMiddleware.hasRequestId}/${requestId} ` })
+                }
+
+                return await cacheMiddleware.cache[requestId][key]
+            }
+        }
+        else {
```
Comment (pahaz) [inline packages/keystone/cache.js:137]:
```suggestion
       } else {
```

**Review by pahaz — APPROVED** [review]

File: packages/keystone/cache.js (hunk 4 above)
Comment (toplenboren) [reply packages/keystone/cache.js:107]: thats just a quick hack ))

File: packages/keystone/cache.js (hunk 6 above)
Comment (toplenboren) [reply packages/keystone/cache.js:50]: ?

File: packages/keystone/cache.js (hunk 7 above)
Comment (toplenboren) [reply packages/keystone/cache.js]: Done

File: packages/keystone/cache.js (hunk 5 above)
Comment (toplenboren) [reply packages/keystone/cache.js]: Done

File: packages/keystone/cache.js (hunk 3 above)
Comment (toplenboren) [reply packages/keystone/cache.js]: Done

File: packages/keystone/cache.js (hunk 6 above)
Comment (toplenboren) [reply packages/keystone/cache.js:50]: Because you also need to transfer cache itself

Comment (toplenboren) [reply packages/keystone/cache.js:50]: So, cache and state is located in the same class

I can make a `typeof` assertion in initCache function, but dont really know if it is going to help. This script is small

File: packages/keystone/cache.js [hunk 8]
```diff
@@ -61,68 +79,98 @@ const getRequestIdFromContext = (context) => {
...
             }
 
             // Drop the key, if the operation type is mutation
             const operationType = get(info, ['operation', 'operation'])
-            if (operationType !== 'query' && get(cache, [requestId, key])) {
-                delete cache[requestId][key]
+            if (operationType !== 'query' && get(cacheMiddleware.cache, [requestId, key])) {
+
+                if (ENABLE_CACHE_LOGGING) {
+                    logger.info(`DELETE: ${requestId}\r\n${gqlName} ${JSON.stringify(args)}`)
```
Comment (sitozzz) [inline packages/keystone/cache.js:117]:
```suggestion
                    logger.info({ msg:`DELETE: ${requestId}\r\n${gqlName}`, args})
```

**Pull request merged**